metagpt/dynamic_sop.py

```diff
@@ -247,59 +246,72 @@ def get_all_agents(self, list_of_agents):
+                    "subtask_description": str,       // <Task Subsection or Requirement>
+                    "agent": str,                     // <Agent Name>
+                    "skill": str,                     // <Agent Skill>
+                    "actions": list[str],             // <Agent Actions>
+                    "watch_items": list[str],         // <Agent Watch Items>
+                    "trigger": str,                   // <Action that triggers this agent>
+                },
+                ...
+            ]
+            ```
+            """
+
+            prompt = Prompt(f"""
+            Your task is to assign agents to a project based on their skills, actions, and watch items.
+            Agents are triggered only when their watch items are updated by the actions of the previous agent. Ensure that all necessary roles are covered in the correct logical sequence, and avoid including agents that are not relevant to the task.
+    
+            You are given:
+    
+            A specific task and domain.
+            A list of agents, their skills, actions, and watch items.
+            Instructions:
+                * Identify and match agents to the task based on their skills, actions, and watch items.
+                * Only select agents whose watch items are triggered by the actions of a previous agent in the sequence.
+                * If no further agent is triggered, stop assigning more agents to the task. This ensures that only the necessary agents are involved.
+                * Ensure that no unnecessary agents are included and that agents are logically arranged based on their dependencies.
+            Selection Criteria:
+                * An agent is only included if their watch items are triggered by the actions of the previous agent.
+                * Skip agents if their actions and watch items are not relevant to the task or if no other agent triggers them.
+                * If a task can be completed by fewer agents (e.g., only the ProductManager for writing a PRD), include only the relevant agent(s).
+            
+            Given Task: {idea}
+            Domain: {domain}
+            Agents: {self.agents}
+    
+            Output Format: format the output in array of json objects, for example
+            {json_example}
+
+            Note:
+            Agent: Represents the selected agent for the task.
+            Skill: The agent's main competency relevant to the task.
+            Actions: The tasks performed by this agent.
+            Watch Items: Items that, when updated by another agent’s actions, trigger this agent’s tasks.
+            Trigger: The action from the previous agent that triggers this agent.
+            """)
+        else:
+            prompt = idea
+        query = prompt.adjusted_prompt if prompt.adjusted_prompt else prompt.prompt
+        agents_response = await self.llm.aask(query, stream=True)
```

Comment:
we can put stream=False, this option prints the llm response to console. we can switch it off